### ISSUE TITLE
CI: Add code coverage for Python code

### ISSUE THREAD

**Issue by seberg** [numpy/numpy#19853]
It would be nice to have python code coverage as part of CI (same as C code coverage).  Unlike the C code coverage, it should be even easier to get very good coverage (since impossible to test error paths are fairly common in C).

I think it would be generally good.  One thing where I am noticing it now is that `loadtxt` is terrible test coverage, and code coverage of the existing code would also help with replacing that code (I can run that local for that concrete use-case).  It might also identify some sprint-able/LGTM style small contributions.

Plus, I think it might simplify review a bit: Instead of asking to make sure that error paths are covered, we can ask to make sure code coverage annotations are gone, giving a clearer picture of how many tests are missing (and why that is not good).

**Comment by BvB93** [comment]
Codecov currently has a github actions plugin (https://github.com/codecov/codecov-action); it might be worth using.

**Comment by ganesh-k13** [comment]
I have experimented on @BvB93 GitHub action and it seems to provide a lot of features. Here is an example of python coverage: https://github.com/ganeshkathiresan/CI-19853-NumPy-Codecov/pull/2

I have integrated the badge feature and a PR comment feature. There are also stuff like minimum %age and lots more. It looks like this:

![Screenshot_20210918_123855](https://user-images.githubusercontent.com/20969920/133879840-062ea0af-18be-406e-99fe-1ee1737f0c68.png)

Let me know if this is what you were looking for.

**Comment by BvB93** [comment]
The codecov comments are currently disabled (per [.codecov.yml](https://github.com/numpy/numpy/blob/main/.codecov.yml), but I think the inline warnings for untested codepaths should still appear, right?
![Untitled](https://user-images.githubusercontent.com/43369155/133894247-ef49d06c-d523-45a3-9711-dac6bb334773.png)

**Comment by seberg** [comment]
Yeah, I assume the settings we use for C just translate and we mainly have to set up the a runner for python coverage that uploads it?  I personally don't see much use in the comments and they can feel spammy, neither do we need the badge (we used to have it for C even).  Basically, we will have the annotations, and if the codecov reports show some things to look at, we can make an issue and link to the website (which also has the full reports).

**Comment by mattip** [comment]
Github's upload action can now upload a (single) html page (no separate formatting css or js files). It might be nice to have a `spin coverage` command that would run `pytest --coverage` and upload the results, extra points for getting c level gcov as well
